ScoutSuite/providers/aws/resources/ec2/instances.py

```diff
@@ -28,3 +30,34 @@ class EC2Instances(AWSResources):
             instance['network_interfaces'][eni['NetworkInterfaceId']] = nic
 
         return id, instance
+
+    @staticmethod
+    def _identify_user_data_secrets(user_data):
+        """
+        Parses EC2 user data in order to identify secrets and credentials..
+        """
+        secrets = {}
+
+        if user_data:
+            aws_access_key_regex = re.compile('AKIA[0-9A-Z]{16}')
+            aws_secret_access_key_regex = re.compile('[0-9a-zA-Z/+]{40}')
+            rsa_private_key_regex = re.compile('(-----BEGIN RSA PRIVATE KEY-----(?s).+?-----END .+?-----)')
+            keywords = ['password', 'secret', 'aws_access_key_id', 'aws_secret_access_key', 'aws_session_token']
+
+            aws_access_key_list = aws_access_key_regex.findall(user_data)
+            if aws_access_key_list:
+                secrets['AWS Access Key IDs'] = aws_access_key_list
+            aws_secret_access_key_list = aws_secret_access_key_regex.findall(user_data)
+            if aws_secret_access_key_list:
+                secrets['AWS Secret Access Keys'] = aws_secret_access_key_list
+            rsa_private_key_list = rsa_private_key_regex.findall(user_data)
+            if rsa_private_key_list:
+                secrets['Private Keys'] = rsa_private_key_list
+            word_list = []
+            for word in keywords:
+                if word in user_data.lower():
+                    word_list.append(word)
+            if word_list:
+                secrets['Flagged Words'] = word_list
+
+        return secrets
```

Comment:
Do you think it would be a good idea to make those regexes and forbidden keywords arguments in the rule, and do the processing directly in the rule?

Reply:
I could, but then I wouldn't be able to show what's being flagged within the user data or why (I think?). As user data can be a lot of content, if you just say "hey I found something, figure out what" it wouldn't be very helpful. 

The current view lists the occurrences identified, what they were identified to be (AWS credentials, RSA keys, suspicious keywords, etc.).

Reply:
Fair enough!